consensus.go

```diff
@@ -0,0 +1,220 @@
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	choice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	ctx := wallet.shutdownContext()
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, choice)
+	if err != nil {
+		return err
+	}
+
+	// If ticket hash is provided, then set vote choice for the selected ticket.
+	if ticketHash != nil {
+		vspTicketInfo, err := wal.VSPTicketInfo(ctx, ticketHash)
+		if err != nil {
+			return err
+		}
+		// Update the vote choice for the ticket with the vsp
+		vspClient, err := wallet.VSPClient(vspTicketInfo.Host, vspTicketInfo.PubKey)
+		if err != nil {
+			return err
+		}
+		err = vspClient.SetVoteChoice(ctx, ticketHash, choice)
+		return err
+	}
+
+	vspClient, err := wallet.VSPClient(vspHost, vspPubKey)
+	if err != nil {
+		return err
+	}
+
+	// Ticket hash wasn't provided, therefore set vote choice for all tickets
+	// belonging to the selected wallet and controlled by the specififed VSP
+	var firstErr error
+	vspClient.ForUnspentUnexpiredTickets(ctx, func(hash *chainhash.Hash) error {
+		// Never return errors here, so all tickets are tried.
+		// The first error will be returned to the user.
+		err := vspClient.SetVoteChoice(ctx, hash, choice)
+		if err != nil && firstErr == nil {
+			firstErr = err
+		}
+		return nil
+	})
+
+	return firstErr
+}
+
+// GetAllAgendasForWallet returns all agendas through out the various
+// stake versions for the active network and this version of the software,
+// and all agendas defined by it for a selected wallet.
+func (wallet *Wallet) GetAllAgendasForWallet(hash string, newestFirst bool) (uint32, []*Agenda, error) {
+	if wallet == nil {
+		return 0, nil, errors.New(ErrNotExist)
+	}
+	wal := wallet.Internal()
+
+	version, deployments := getAllAgendas(wallet.chainParams)
+	agendas := make([]*Agenda, len(deployments))
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return 0, nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	voteChoicesResult := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]AgendaVoteChoice, len(deployments)),
+	}
+
+	ctx := wallet.shutdownContext()
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	for i := range choices {
+		agenda := agenda(deployments, choices[i].AgendaID)
+		voteChoicesResult.Choices[i] = AgendaVoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agenda.Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for _, choice := range agenda.Vote.Choices {
+			if choices[i].ChoiceID == choice.Id {
+				voteChoicesResult.Choices[i].ChoiceDescription = choice.Description
+				break
+			}
+		}
+	}
+
+	for i := range deployments {
```

Comment:
I'm quite convinced these 2 for loops can be merged. The first one only populates a temporary field that this one accesses.